core/src/net_connection.rs

```diff
@@ -0,0 +1,534 @@
+use crate::avm2::object::{
+    NetConnectionObject as Avm2NetConnectionObject, ResponderObject as Avm2ResponderObject,
+};
+use crate::avm2::{Activation as Avm2Activation, Avm2, EventObject as Avm2EventObject};
+use crate::backend::navigator::{NavigatorBackend, OwnedFuture, Request};
+use crate::context::UpdateContext;
+use crate::loader::Error;
+use crate::string::AvmString;
+use crate::Player;
+use flash_lso::packet::{Header, Message, Packet};
+use flash_lso::types::{AMFVersion, Value as AmfValue};
+use gc_arena::{Collect, DynamicRoot, Rootable};
+use generational_arena::{Arena, Index};
+use std::fmt::{Debug, Formatter};
+use std::rc::Rc;
+use std::sync::{Mutex, Weak};
+
+pub type NetConnectionHandle = Index;
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum ResponderCallback {
+    Result,
+    Status,
+}
+
+#[derive(Clone)]
+pub enum ResponderHandle {
+    Avm2(DynamicRoot<Rootable![Avm2ResponderObject<'_>]>),
+}
+
+impl Debug for ResponderHandle {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ResponderHandle::Avm2(_) => write!(f, "ResponderHandle::Avm2"),
+        }
+    }
+}
+
+impl ResponderHandle {
+    pub fn call(
+        &self,
+        context: &mut UpdateContext<'_, '_>,
+        callback: ResponderCallback,
+        message: Rc<AmfValue>,
+    ) {
+        match self {
+            ResponderHandle::Avm2(handle) => {
+                let object = context.dynamic_root.fetch(handle);
+                if let Err(e) = object.send_callback(context, callback, &message) {
+                    tracing::error!("Unhandled error sending {callback:?} callback: {e}");
+                }
+            }
+        }
+    }
+}
+
+#[derive(Copy, Clone, Collect)]
+#[collect(no_drop)]
+pub enum NetConnectionObject<'gc> {
+    Avm2(Avm2NetConnectionObject<'gc>),
+}
+
+impl<'gc> NetConnectionObject<'gc> {
+    pub fn set_handle(&self, handle: Option<NetConnectionHandle>) -> Option<NetConnectionHandle> {
+        match self {
+            NetConnectionObject::Avm2(object) => object.set_handle(handle),
+        }
+    }
+}
+
+impl<'gc> From<Avm2NetConnectionObject<'gc>> for NetConnectionObject<'gc> {
+    fn from(value: Avm2NetConnectionObject<'gc>) -> Self {
+        NetConnectionObject::Avm2(value)
+    }
+}
+
+/// Manages the collection of NetConnections.
+pub struct NetConnections<'gc> {
+    connections: Arena<NetConnection<'gc>>,
+}
+
+unsafe impl<'gc> Collect for NetConnections<'gc> {
+    fn trace(&self, cc: &gc_arena::Collection) {
+        for (_, connection) in self.connections.iter() {
+            connection.trace(cc)
+        }
+    }
+}
+
+impl<'gc> Default for NetConnections<'gc> {
+    fn default() -> Self {
+        Self {
+            connections: Arena::new(),
+        }
+    }
+}
+
+impl<'gc> NetConnections<'gc> {
```

Comment:
I don't know. In the future there'd be connecting to rtmp, connecting to rtmpt, etc. "Flash remoting" and "local" are the names of the protocol, and just "connect_remote" is a bit confusing when there's many remotes it could connect to.